Add Github Action for Tests

eitanturok opened the pull request:
In this PR, I
1. Migrated 400+ tests from absl-py to pytest
2. Add `pyproject.toml` to handle us importing `torch` in `setup.py`
3. Add packages to the `setup.py` dev requirements for testing
4. In `setup.py`, use up to date versions of `grouped_gemm` and `stk` packages that properly handle build isolation (@tgale96 merged my PRs into those libraries.)
5. Update `.gitignore`, based on `Composer`'s `.gitignore`
6. Update `Makefile` to be used with testing, based on `Composer`'s `Makefile`
7. Add GA for testing on 1 and 8 GPUs on every push using `ci-testing` repo, based on `Composer`'s GA

Notes:
1. I kept the original absl-py test files for all the tests that I migrated to pytest. I will delete these in a separate PR, once I've finished migrating all tests to pytest. This makes it so I can make sure all the absl-py and the pytest tests pass at the same time.
2. There are certain tests that run on multiple GPUs or are launched using a bash script. I have not yet migrated these to pytest but will do so in a subsequent PR.

mvpatel2000 commented on setup.py:
```diff
@@ -28,19 +29,28 @@
+    'pytest-cov>=4,<5',
+    'pytest>=7.2.1,<8',
+    'pre-commit>=3.4.0,<4',
+    'mosaicml>=0.22.0',
```
why do we need composer as a dependency?

eitanturok replied:
If we run tests on multiple GPUs, we call the Makefile target `test-dist` which calls `composer.cli.launcher`. 

Turbo also requires `composer` as a dependency for the same reason.

b-chu replied:
we shouldn't have composer as a dependency. Maybe something like this? https://github.com/mosaicml/streaming/blob/main/setup.py#L130

mvpatel2000 reviewed (commented):
Looks good to me as initial PR.

@tgale96  any other tests you want to include / would be helpful?

eitanturok commented on setup.py:
```diff
@@ -28,19 +29,28 @@
+    'numpy>=1.21.5,<2.1.0',
+    'torch>=2.3.0,<2.4',
+    'triton>=2.1.0',
+    'stanford-stk @ git+https://git@github.com/stanford-futuredata/stk.git@a1ddf98466730b88a2988860a9d8000fd1833301',
```
@tgale96 needs to cut a new release. We should talk to him about this. In the meantime though, I just pinned to the last commit. I think we should merge with this and update when @tgale96 makes the new release.

eitanturok commented on setup.py:
```diff
@@ -28,19 +29,28 @@
 
 extra_deps["gg"] = [
-    "grouped_gemm==0.1.4",
+    'grouped_gemm @ git+https://git@github.com/tgale96/grouped_gemm.git@66c7195e35e8c4f22fa6a014037ef511bfa397cb',
```
@tgale96 needs to cut a new release. The last pinned release breaks because of the build in isolation error. Again, we should talk to Trevor about this.

In the meantime though, I just pinned to the last commit. I think we should merge with this and update when @tgale96 makes the new release.

eitanturok commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,46 @@
+  push:
+    branches:
+    - main
+    - dev
```
Removed.

eitanturok commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,46 @@
+  pull_request_target:
+    branches:
+    - main
+    - dev
```
Removed.

eitanturok commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,46 @@
+        include:
+        - name: "python3.11-pytorch2.3-gpus1"
+          gpu_num: 1
+          container: mosaicml/pytorch:2.3.0_cu121-python3.11-ubuntu20.04
```
Updated.

eitanturok commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,46 @@
+          container: mosaicml/pytorch:2.3.0_cu121-python3.11-ubuntu20.04
+        - name: "python3.11-pytorch2.3-gpus8"
+          gpu_num: 8
+          container: mosaicml/pytorch:2.3.0_cu121-python3.11-ubuntu20.04
```
Updated.

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,44 @@
+    branches:
+    - main
+    - release/**
+    - eitan-tests
```
?

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,44 @@
+  cancel-in-progress: ${{ github.ref != 'refs/heads/main' && github.ref != 'refs/heads/dev' }}
+jobs:
+  gpu-pytests:
+    uses: mosaicml/ci-testing/.github/workflows/pytest-gpu.yaml@v0.0.9
```
outdated

b-chu replied:
why did you resolve this?

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,44 @@
+          gpu_num: 1
+          container: mosaicml/pytorch:2.3.1_cu121-python3.11-ubuntu20.04
+        - name: "python3.11-pytorch2.3.1-gpus8"
+          gpu_num: 8
```
why 8?

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,44 @@
+          gpu_num: 8
+          container: mosaicml/pytorch:2.3.1_cu121-python3.11-ubuntu20.04
+    name: ${{ matrix.name }}
+    if: github.repository_owner == 'eitanturok'  # todo: replace with `databricks`
```
^

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,44 @@
+    if: github.repository_owner == 'eitanturok'  # todo: replace with `databricks`
+    with:
+      container: ${{ matrix.container }}
+      git_repo: eitanturok/megablocks  # todo: replace with `databricks/megablocks`
```
^

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,44 @@
+      composer_package_name: mosaicml  # Required as Composer is built from source
+      gpu_num: ${{ matrix.gpu_num }}
+    secrets:
+      mcloud-api-key: ${{ secrets.MCLOUD_API_KEY }}
```
do you have secrets set up?

b-chu replied:
why is this resolved?

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,44 @@
+    branches:
+    - main
+    - release/*
+    - eitan-tests
```
?

b-chu commented on setup.py:
```diff
@@ -28,19 +29,28 @@
-extra_deps["dev"] = [
-    "absl-py",
+extra_deps['dev'] = [
+    'absl-py',
```
Why do we still need this?

b-chu commented on tests/test_smoketest.py:
```diff
@@ -0,0 +1,12 @@
+
+# This very simple test is just to use the above imports, which check and make sure we can import all the top-level
+# modules from Megablocks. This is mainly useful for checking that we have correctly conditionally imported all optional
+# dependencies.
```
> correctly conditionally imported all optional dependencies

What does this mean?

eitanturok replied:
This test was taken from Composer. Essentially it is a simple test to make sure we import everything correctly.

mvpatel2000 commented on setup.py:
```diff
@@ -28,22 +29,34 @@
 ]
 
-extra_deps['all'] = set(dep for deps in extra_deps.values() for dep in deps)
+extra_deps['all'] = list(set(dep for deps in extra_deps.values() for dep in deps))
```
all should exclude `testing`. See in foundry how we exclude some options

mvpatel2000 reviewed (commented):
LGTM besides one last comment. Asking @b-chu  to also approve

eitanturok commented:
Updates:
1. Thanks for the comments. I addressed all of them. 
2. I've migrated 450+ tests from absl-py to pytest, including low level tests for individual ops and higher-level tests that ensure MoEs are implemented properly across various sparsity, memory-optimization, and dropless settings.
3. The Github Action for testing should work. For reference, the Github Actions were working in my fork [here](https://github.com/eitanturok/megablocks/actions/runs/10165283474).
4. `parallelism_test.py` required 8 GPUs. I've updated it to only use 2 GPUs. We now only run testing jobs on 1 and 2 GPUs.
5. I've since added a `conftest.py` file and various autouse fixtures to make it easier to test with distributed inference and add bells and whistles. 

To do:
3. Do we want to cut a new tag for stk and grouped_gemm for this release?
2. Should I delete the old test files that were written in absl-py (and remove the absl-py dpeendency)?

mvpatel2000 commented:
> 3. Do we want to cut a new tag for stk and grouped_gemm for this release? 

Let's do in follow-on PR

> 4. 2. Should I delete the old test files that were written in absl-py (and remove the absl-py dpeendency)?

Yes I think so.

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,40 @@
+  workflow_dispatch:
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/main' && github.ref != 'refs/heads/dev' }}
```
there's no dev

b-chu commented on .github/workflows/pr-gpu.yaml:
```diff
@@ -0,0 +1,40 @@
+      fail-fast: false
+      matrix:
+        include:
+        - name: "python3.11-pytorch2.3.1-gpus1"
```
3.11 here but 3.9 below

b-chu reviewed (commented):
Eitan, you shouldn't resolve conversations unless you address the comment or answer the question

Have you tested setup.py? Are all tests supposed to be GPU tests?

eitanturok commented:
> Eitan, you shouldn't resolve conversations unless you address the comment or answer the question
> 
> Have you tested setup.py? Are all tests supposed to be GPU tests?

Sorry, I shouldn't have resolved those conversations. I've since addressed those issues that I resolved before. (We are now using v0.1.0 of ci-testing.)

I do not have explicit tests for the setup.py. But when we install MegaBlocks to run all of our tests, we implicitly need the setup.py to work correctly. Are there any specific tests you recommend adding for the setup.py?

And yes, all tests are currently GPU tests. We are testing ops, which are wrappers around low level kernels; these kernels only work on GPUs. We are also testing different MoE configurations where we need to perform inference; again, this is done on GPUs because it is inference. There is not much more in MegaBlocks beyond this and so there are not many non-GPU tests to run. Happy to add non-GPU tests if you have any in mind.

b-chu reviewed (commented):
<img width="801" alt="image" src="https://github.com/user-attachments/assets/a1cb8d60-6871-4041-ac08-b2c1283d38a8">

eitanturok commented:
> Have you manually run setup and GPU tests?

Yes.

b-chu reviewed (approved):
Thanks for the PR!
